implementations/rust/ockam/ockam_executor/src/channel.rs

```diff
@@ -0,0 +1,151 @@
+    queue: Queue<T>,
+
+    /// Notifies all tasks listening for the receiver being dropped
+    wake_sender: AtomicWaker,
+
+    /// Receiver waker. Notified when a value is pushed into the channel.
+    wake_receiver: AtomicWaker,
+
+    /// Tracks the number of outstanding sender handles.
+    ///
+    /// When this drops to zero, the send half of the channel is closed.
+    sender_count: AtomicUsize,
+
+    /// Set to true when the send half of the channel is closed.
+    is_sender_closed: AtomicBool,
+}
+
+pub struct Sender<T>(Arc<Inner<T>>);
+
+impl<T: core::fmt::Debug> Sender<T> {
+    pub async fn send(&self, value: T) -> Result<(), error::SendError<T>> {
+        let mut value = Some(value);
+        poll_fn(|context| {
+            match self.0.queue.enqueue(value.take().unwrap()) {
+                Ok(()) => {
+                    self.0.wake_receiver.wake();
+                    Poll::Ready(Ok(()))
+                }
+                Err(_) => {
+                    // queue is full
+                    {
+                        self.0.is_sender_closed.swap(true, Ordering::AcqRel);
+                        self.0.wake_receiver.wake();
+                    }
+                    self.0.wake_sender.register(&context.waker());
+                    Poll::Pending
+                }
+            }
+        })
+        .await
+    }
+
+    pub async fn closed(&self) {
+        unimplemented!();
+    }
+}
+
+impl<T> Clone for Sender<T> {
+    fn clone(&self) -> Self {
+        self.0.sender_count.fetch_add(1, Ordering::Relaxed);
+        Sender(self.0.clone())
+    }
+}
+
+impl<T> Drop for Sender<T> {
+    fn drop(&mut self) {
+        let sender_count = self.0.sender_count.fetch_sub(1, Ordering::AcqRel);
+        if sender_count != 1 {
+            return;
+        }
+
+        // close the list
+        self.0.is_sender_closed.swap(true, Ordering::AcqRel);
+
+        // notify the receiver
+        self.0.wake_receiver.wake();
+    }
+}
+
+impl<T> core::fmt::Debug for Sender<T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(f, "Sender")
+    }
+}
+
+pub struct Receiver<T>(Arc<Inner<T>>);
+
+impl<T: core::fmt::Debug> Receiver<T> {
+    pub async fn recv(&mut self) -> Option<T> {
+        let result = poll_fn(|context| match self.0.queue.dequeue() {
+            Some(value) => {
+                self.0.wake_sender.wake();
+                Poll::Ready(Some(value))
+            }
+            None => {
+                self.0.wake_receiver.register(&context.waker());
+                if self.0.is_sender_closed.load(Ordering::Relaxed) {
+                    Poll::Ready(None)
+                } else {
+                    Poll::Pending
+                }
+            }
+        })
+        .await;
+        result
+    }
+}
+
+impl<T> core::fmt::Debug for Receiver<T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
+        write!(f, "[Receiver]")
```

Comment:
If we're concerned about code size, `f.write_str` might be better.

It might not matter without format arguments though, I don't remember.

Reply:
Good to know!